File: src/core/data_processor.py
#!/usr/bin/env python3
"""Data processor with TI AWR1843BOOST parser"""
import threading
import queue
import time
import sys
import os
from typing import Optional, Dict, List

# Add parsers to path
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'parsers'))
from awr1843_parser import AWR1843Parser

class DataProcessor(threading.Thread):
    def __init__(self, buffers: List, output_queue: queue.Queue):
        super().__init__(daemon=True)
        self.buffers = buffers
        self.output_queue = output_queue
        self.running = True
        self.parser = AWR1843Parser()
        
    def run(self):
        while self.running:
            for buffer in self.buffers:
                frames = buffer.get_all()
                for frame in frames:
                    # Parse AWR1843 frame - returns list of detected objects
                    detected_objects = self.parser.parse_frame(frame.raw_data, frame.radar_id)
                    
                    # Send each parsed data item as separate frame
                    for data_item in detected_objects:
                        # Create new frame for each data item
                        new_frame = type('RadarFrame', (), {
                            'radar_id': frame.radar_id,
                            'timestamp': frame.timestamp,
                            'raw_data': frame.raw_data,
                            'parsed_data': data_item
                        })()
                        try:
                            self.output_queue.put_nowait(new_frame)
                        except queue.Full:
                            pass
                            
            time.sleep(0.0005)  # 2000Hz processing loop
    
    def stop(self):
        self.running = False

File: cpp/src/core/data_processor.cpp
#include "core/data_processor.h"
#include "parsers/awr1843_parser.h"
#include <chrono>
#include <iostream>

DataProcessor::DataProcessor(std::vector<std::shared_ptr<RingBuffer<RadarFrame>>>& buffers,
                           std::queue<RadarFrame>& output_queue)
    : buffers_(buffers), output_queue_(output_queue), running_(false) {
    parser_ = std::make_unique<AWR1843Parser>();
}

DataProcessor::~DataProcessor() {
    stop();
}

void DataProcessor::start() {
    running_ = true;
    worker_thread_ = std::thread(&DataProcessor::run, this);
    std::cout << "[DataProcessor] Started with optimized AWR1843 parser" << std::endl;
}

void DataProcessor::stop() {
    running_ = false;
    if (worker_thread_.joinable()) {
        worker_thread_.join();
    }
    std::cout << "[DataProcessor] Stopped" << std::endl;
}

void DataProcessor::run() {
    while (running_) {
        for (auto& buffer : buffers_) {
            auto frames = buffer->get_all();
            for (const auto& frame : frames) {
                // Parse AWR1843 frame - returns vector of detected objects
                auto detected_objects = parser_->parse_frame(frame.raw_data, frame.radar_id);
                
                // Create separate frame for each detected object
                for (const auto& obj : detected_objects) {
                    RadarFrame new_frame;
                    new_frame.radar_id = frame.radar_id;
                    new_frame.timestamp = frame.timestamp;
                    new_frame.raw_data = frame.raw_data;
                    new_frame.parsed_data = {obj};
                    
                    // Non-blocking queue insert
                    if (output_queue_.size() < 5000) {
                        output_queue_.push(new_frame);
                    }
                }
            }
        }
        
        // High-frequency processing loop (2000Hz)
        std::this_thread::sleep_for(std::chrono::microseconds(500));
    }
}